### PR TITLE
Cleanup XSPEC interface

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

A small number of XSPEC model parameter values have been changed to match the default frozen state of XSPEC version 12.13.1.

# Details

The user-visible change is that some parameters have had their frozen status changed to match XSPEC. It's actually one of the smallest changes, but I needed to update `scripts/check_xspec_update.py` to actually check for the `frozen` status of parameters, and then things snow-balled out of control as I started adding support for the 50 new models in XSPEC 12.14.0...

Internal re-arrangement of the code linking to the XSPEC API to better match the ordering used by XSPEC and to make it easier to update to newer versions of XSPEC. There is also some clean up of the `norm` parameter for which we have had values that differ on the `hard_max` value (10^24 vs ~3 10^38); we now use a single routine to generate this so that all additive models use the same value. Two notes about this

1. we could do this for other parameters, but that would be noise and better done when needed (e.g. as is the case in the follow-up XSPEC 12.14.0 code)
2. originally I was not going to do this, but I ended up changing the `XSAdditiveModel` class to automatically add the `norm` parameter for the user, as long as they had not specified it themselves. This reduces slightly the boiler plate code needed, but still allows a user control over the `norm` parameter if they want (there are several models whose documentation points out the `norm` should be frozen but there is no way to know this from the `model.dat` file)

This adds a new script, `scripts/update_xspec_functions.py` which will create the basic code needed to link to the XSPEC model functions. For example

```
% ./scripts/update_xspec_functions.py ~/local/heasoft-6.32/spectral/manager/model.dat
  XSPECMODELFCT_C_NORM(C_agauss, 3),               // XSagauss
  XSPECMODELFCT_NORM(agnsed, 16),                  // XSagnsed
  XSPECMODELFCT_NORM(agnslim, 15),                 // XSagnslim
  XSPECMODELFCT_C_NORM(C_apec, 4),                 // XSapec
  XSPECMODELFCT_C_NORM(C_bapec, 5),                // XSbapec
  XSPECMODELFCT_C_NORM(C_btapec, 6),               // XSbtapec
  XSPECMODELFCT_NORM(xsblbd, 2),                   // XSbbody
...
  XSPECMODELFCT_CON(C_simpl, 3),                   // XSsimpl
  XSPECMODELFCT_CON_F77(thcompf, 4),               // XSthcomp
  XSPECMODELFCT_CON(C_vashift, 1),                 // XSvashift
  XSPECMODELFCT_CON(C_vmshift, 1),                 // XSvmshift
  XSPECMODELFCT_CON(C_xilconv, 6),                 // XSxilconv
  XSPECMODELFCT_CON(C_zashift, 1),                 // XSzashift
  XSPECMODELFCT_CON(C_zmshift, 1),                 // XSzmshift

  XSPECMODELFCT_C_NORM(beckerwolff, 13),           // XSbwcycl
```

It allows several things

a) comparing versions (although we have other ways to do this)

``` 
% ./scripts/update_xspec_functions.py ~/local/heasoft-6.32/spectral/manager/model.dat  > a.lis
% ./scripts/update_xspec_functions.py ~/local/heasoft-6.33.1/spectral/manager/model.dat  > b.lis
% diff -u a.lis b.lis
--- a.lis	2024-04-09 15:26:54.906725504 -0400
+++ b.lis	2024-04-09 15:27:05.314708802 -0400
@@ -3,6 +3,17 @@
   XSPECMODELFCT_NORM(agnslim, 15),                 // XSagnslim
   XSPECMODELFCT_C_NORM(C_apec, 4),                 // XSapec
   XSPECMODELFCT_C_NORM(C_bapec, 5),                // XSbapec
+  XSPECMODELFCT_C_NORM(C_bcempow, 8),              // XSbcempow
+  XSPECMODELFCT_C_NORM(C_bcheb6, 12),              // XSbcheb6
+  XSPECMODELFCT_C_NORM(C_bcie, 6),                 // XSbcie
+  XSPECMODELFCT_C_NORM(C_bcoolflow, 7),            // XSbcoolflow
+  XSPECMODELFCT_C_NORM(C_bcph, 6),                 // XSbcph
+  XSPECMODELFCT_C_NORM(C_bequil, 5),               // XSbequil
+  XSPECMODELFCT_C_NORM(C_bexpcheb6, 12),           // XSbexpcheb6
+  XSPECMODELFCT_C_NORM(C_bgaussDem, 8),            // XSbgadem
+  XSPECMODELFCT_C_NORM(C_bgnei, 7),                // XSbgnei
+  XSPECMODELFCT_C_NORM(C_bnei, 6),                 // XSbnei
+  XSPECMODELFCT_C_NORM(C_bsnapec, 8),              // XSbsnapec
   XSPECMODELFCT_C_NORM(C_btapec, 6),               // XSbtapec
   XSPECMODELFCT_NORM(xsblbd, 2),                   // XSbbody
   XSPECMODELFCT_NORM(xsbbrd, 2),                   // XSbbodyrad
@@ -11,22 +22,51 @@
   XSPECMODELFCT_C_NORM(C_brokenPowerLaw, 4),       // XSbknpower
   XSPECMODELFCT_C_NORM(C_broken2PowerLaw, 6),      // XSbkn2pow
   XSPECMODELFCT_NORM(xsbmc, 4),                    // XSbmc
+  XSPECMODELFCT_C_NORM(C_bnpshock, 8),             // XSbnpshock
+  XSPECMODELFCT_C_NORM(C_bpshock, 7),              // XSbpshock
   XSPECMODELFCT_NORM(xsbrms, 2),                   // XSbremss
   XSPECMODELFCT_C_NORM(C_brnei, 7),                // XSbrnei
+  XSPECMODELFCT_C_NORM(C_bsedov, 7),               // XSbsedov
   XSPECMODELFCT_C_NORM(C_bvapec, 17),              // XSbvapec
+  XSPECMODELFCT_C_NORM(C_bvcempow, 21),            // XSbvcempow
+  XSPECMODELFCT_C_NORM(C_bvcheb6, 25),             // XSbvcheb6
+  XSPECMODELFCT_C_NORM(C_bvcie, 18),               // XSbvcie
+  XSPECMODELFCT_C_NORM(C_bvcoolflow, 20),          // XSbvcoolflow
+  XSPECMODELFCT_C_NORM(C_bvcph, 19),               // XSbvcph
+  XSPECMODELFCT_C_NORM(C_bvequil, 16),             // XSbvequil
+  XSPECMODELFCT_C_NORM(C_bvexpcheb6, 25),          // XSbvexpcheb6
+  XSPECMODELFCT_C_NORM(C_bvgaussDem, 21),          // XSbvgadem
+  XSPECMODELFCT_C_NORM(C_bvgnei, 19),              // XSbvgnei
+  XSPECMODELFCT_C_NORM(C_bvnei, 18),               // XSbvnei
+  XSPECMODELFCT_C_NORM(C_bvnpshock, 20),           // XSbvnpshock
+  XSPECMODELFCT_C_NORM(C_bvpshock, 19),            // XSbvpshock
   XSPECMODELFCT_C_NORM(C_bvrnei, 19),              // XSbvrnei
+  XSPECMODELFCT_C_NORM(C_bvsedov, 19),             // XSbvsedov
   XSPECMODELFCT_C_NORM(C_bvtapec, 18),             // XSbvtapec
   XSPECMODELFCT_C_NORM(C_bvvapec, 34),             // XSbvvapec
+  XSPECMODELFCT_C_NORM(C_bvvcie, 35),              // XSbvvcie
+  XSPECMODELFCT_C_NORM(C_bvvgaussDem, 37),         // XSbvvgadem
+  XSPECMODELFCT_C_NORM(C_bvvgnei, 36),             // XSbvvgnei
+  XSPECMODELFCT_C_NORM(C_bvvnei, 35),              // XSbvvnei
+  XSPECMODELFCT_C_NORM(C_bvvnpshock, 37),          // XSbvvnpshock
+  XSPECMODELFCT_C_NORM(C_bvvpshock, 36),           // XSbvvpshock
   XSPECMODELFCT_C_NORM(C_bvvrnei, 36),             // XSbvvrnei
+  XSPECMODELFCT_C_NORM(C_bvvsedov, 36),            // XSbvvsedov
   XSPECMODELFCT_C_NORM(C_bvvtapec, 35),            // XSbvvtapec
+  XSPECMODELFCT_C_NORM(C_bvvwDem, 38),             // XSbvvwdem
+  XSPECMODELFCT_C_NORM(C_bvwDem, 22),              // XSbvwdem
+  XSPECMODELFCT_C_NORM(C_bwDem, 9),                // XSbwdem
   XSPECMODELFCT_C_NORM(C_c6mekl, 11),              // XSc6mekl
   XSPECMODELFCT_C_NORM(C_c6pmekl, 11),             // XSc6pmekl
   XSPECMODELFCT_C_NORM(C_c6pvmkl, 24),             // XSc6pvmkl
   XSPECMODELFCT_C_NORM(C_c6vmekl, 24),             // XSc6vmekl
   XSPECMODELFCT_C_NORM(C_carbatm, 4),              // XScarbatm
-  XSPECMODELFCT_NORM(cemekl, 7),                   // XScemekl
+  XSPECMODELFCT_C_NORM(C_cemMekal, 7),             // XScemekl
+  XSPECMODELFCT_C_NORM(C_cempow, 7),               // XScempow
   XSPECMODELFCT_C_NORM(C_cemVMekal, 20),           // XScevmkl
   XSPECMODELFCT_C_NORM(C_xscflw, 6),               // XScflow
+  XSPECMODELFCT_C_NORM(C_cheb6, 11),               // XScheb6
+  XSPECMODELFCT_C_NORM(C_cie, 5),                  // XScie
   XSPECMODELFCT_NORM(compbb, 4),                   // XScompbb
   XSPECMODELFCT_C_NORM(xscompmag, 9),              // XScompmag
   XSPECMODELFCT_NORM(compls, 3),                   // XScompLS
@@ -35,6 +75,7 @@
   XSPECMODELFCT_C_NORM(xscomptb, 7),               // XScomptb
   XSPECMODELFCT_C_NORM(C_xscompth, 21),            // XScompth
   XSPECMODELFCT_NORM(xstitg, 6),                   // XScompTT
+  XSPECMODELFCT_C_NORM(C_coolflow, 6),             // XScoolflow
   XSPECMODELFCT_C_NORM(C_cph, 5),                  // XScph
   XSPECMODELFCT_C_NORM(C_cplinear, 21),            // XScplinear
   XSPECMODELFCT_C_NORM(C_cutoffPowerLaw, 3),       // XScutoffpl
@@ -46,11 +87,13 @@
   XSPECMODELFCT_NORM(disko, 5),                    // XSdisko
   XSPECMODELFCT_NORM(diskpbb, 3),                  // XSdiskpbb
   XSPECMODELFCT_NORM(xsdiskpn, 3),                 // XSdiskpn
+  XSPECMODELFCT_C_NORM(C_eebremss, 4),             // XSeebremss
   XSPECMODELFCT_NORM(eplogpar, 3),                 // XSeplogpar
   XSPECMODELFCT_C_NORM(C_xseqpair, 21),            // XSeqpair
   XSPECMODELFCT_C_NORM(C_xseqth, 21),              // XSeqtherm
   XSPECMODELFCT_C_NORM(C_equil, 4),                // XSequil
   XSPECMODELFCT_NORM(xsxpdec, 2),                  // XSexpdec
+  XSPECMODELFCT_C_NORM(C_expcheb6, 11),            // XSexpcheb6
   XSPECMODELFCT_NORM(ezdiskbb, 2),                 // XSezdiskbb
   XSPECMODELFCT_C_NORM(C_gaussianLine, 3),         // XSgaussian
   XSPECMODELFCT_C_NORM(C_gaussDem, 7),             // XSgadem
@@ -112,7 +155,12 @@
   XSPECMODELFCT_C_NORM(C_tapec, 5),                // XStapec
   XSPECMODELFCT_C_NORM(C_vapec, 16),               // XSvapec
   XSPECMODELFCT_NORM(xsbrmv, 3),                   // XSvbremss
+  XSPECMODELFCT_C_NORM(C_vcempow, 20),             // XSvcempow
+  XSPECMODELFCT_C_NORM(C_vcheb6, 24),              // XSvcheb6
+  XSPECMODELFCT_C_NORM(C_vcie, 17),                // XSvcie
+  XSPECMODELFCT_C_NORM(C_vcoolflow, 19),           // XSvcoolflow
   XSPECMODELFCT_C_NORM(C_vcph, 18),                // XSvcph
+  XSPECMODELFCT_C_NORM(C_vexpcheb6, 24),           // XSvexpcheb6
   XSPECMODELFCT_C_NORM(C_vequil, 15),              // XSvequil
   XSPECMODELFCT_C_NORM(C_vgaussDem, 20),           // XSvgadem
   XSPECMODELFCT_C_NORM(C_vgnei, 18),               // XSvgnei
@@ -128,6 +176,8 @@
   XSPECMODELFCT_C_NORM(C_vsedov, 18),              // XSvsedov
   XSPECMODELFCT_C_NORM(C_vtapec, 17),              // XSvtapec
   XSPECMODELFCT_C_NORM(C_vvapec, 33),              // XSvvapec
+  XSPECMODELFCT_C_NORM(C_vvcie, 34),               // XSvvcie
+  XSPECMODELFCT_C_NORM(C_vvgaussDem, 36),          // XSvvgadem
   XSPECMODELFCT_C_NORM(C_vvgnei, 35),              // XSvvgnei
   XSPECMODELFCT_C_NORM(C_vvnei, 34),               // XSvvnei
   XSPECMODELFCT_C_NORM(C_vvnpshock, 36),           // XSvvnpshock
```

b) you can use this to add the necessary text to the `_xspec.cc` file now that we have it organised, whereas before we had no order which made checking existing settings a lot harder to do 

All is not perfect, since we have to support symbols that change with the XSPEC version, and we can't easily support this in the script. So there's still some manual work needed, but it will make my life a lot easier when addressing #1987 

In doing this I realized we needed to update the existing code (used manually be external code to support linking to XSPEC user models) to support double-precision multiplicative FORTRAN models. Our existing XSPEC interface already supported this (i.e. already used `XSPECMODELFCT_DBL` rather than `XSPECMODELFCT`) so it's more useful for anyone that uses `scripts/add_xspec_model.py`.

### CODE REVIEW

**Comment by DougBurke** [comment]
Rebased to pick up the codecov fixes in #2000.

**Comment by DougBurke** [comment]
Hmm. Well, we now know that XSgrbjet can still create 0's with XSPEC 12.13.1 (thanks to a failed test), so I guess I can just revert that change! Or, rather, just remove it!

**Review by hamogu — COMMENTED** [review]
I have not checked that the XPEC models are set correctly for each version of XSPEC. If Dough says it's OK and it passes the tests, it's probably OK - there is not much more I can do about it.

May main question is if it's worthwhile to do more of what this PR does with "norm". There are serveral classes of model (e.g. the APEC-type models) that have very repetitive set of of parameters, e.g. abundances for 31 elements that are all identical except for the name. Is it worthwhile to pull that out  into a function or class definition like it's done already for "norm"? Or is it easier and more flexible to keep it as is now and manually update when needed?

File: sherpa/astro/xspec/__init__.py [hunk 1]
```diff
@@ -2627,8 +2652,11 @@ def __init__(self, name='bvapec'):
         self.Ni = XSParameter(name, 'Ni', 1., 0., 1000., 0.0, 1000, frozen=True)
         self.Redshift = XSParameter(name, 'Redshift', 0., -0.999, 10., -0.999, 10, frozen=True)
         self.Velocity = XSParameter(name, 'Velocity', 0., 0., 1.e4, 0.0, 1e4, units='km/s', frozen=True)
-        self.norm = Parameter(name, 'norm', 1.0, 0.0, 1.0e24, 0.0, hugeval)
-        XSAdditiveModel.__init__(self, name, (self.kT, self.He, self.C, self.N, self.O, self.Ne, self.Mg, self.Al, self.Si, self.S, self.Ar, self.Ca, self.Fe, self.Ni, self.Redshift, self.Velocity, self.norm))
+
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/astro/xspec/__init__.py:2655]: This is better discussed after #1997 - which adds some helper functions which just return parameter objects. I did explore a routine more like your `add_abund` (but for single parameters) but there are interesting questions over auto-generating this code (whether as part of #52 or just as part of our `scripts/` package just to help adding new models).

File: scripts/update_xspec_functions.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    # Add in the model name as a comment
+    print(f"{code[0]:50s} // XS{model.name}")
+    return model.modeltype
+
+
+if __name__ == "__main__":
+
+    if len(sys.argv) != 2:
+        sys.stderr.write(f"Usage: {sys.argv[0]} infile\n")
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply scripts/update_xspec_functions.py:94]: These are so simple I just can't be bothered to remember what the latest  @package-for-argument-parsing-du-jour is!

**Review by hamogu — APPROVED** [review]
As I said, just comments because I don't know the order in which you address things in follow-up PRs. Fine with me!

**Pull request merged**